### DISCUSSION TITLE
Question about `strict_types`

### DISCUSSION THREAD

**Discussion by AlwinGarside** [mario-deluna/php-glfw#35] [Q&A] [answered]
Heya, I just discovered this project and I'm very impressed so far.

I have a question about `strict_types`: I noticed `strict_types` is very consistently _not_ enabled in php-glfw and VISU. Don't worry, I don't want to start a typed vs non-typed code discussion here, but I'm very interested whether this is due to performance considerations, or simply due to preference?

If it's the former (performance), I'd really like to learn more. And if it's due to preference, then that's all fine by me.

**Comment by mario-deluna** [comment, marked as answer]
Good point. If anything, the `strict_types` directive could potentially even shave off a few microseconds, as from my understanding, the Zend engine would not need to perform potential implicit casts (though I haven't checked the source to back up that claim).

To be completely honest, I'm not too happy that PHP does type checking at runtime, but I understand why they had to take that route.

When it comes to performance, there are few "micro-optimizations" that actually matter, but it mostly comes down to the following:

- **Use the right data structures for your problem**. 
  Always try to reduce computational complexity. For example, if collision detection is your bottleneck, use an octree, etc.
- **Implement heavily used math functions in C.** 
   For example, matrix multiplications are heavily used in most games. The Mat4 implementation in php-glfw will be an order of magnitude faster than a counterpart implemented in PHP.
- **Avoid getter and setter functions on state objects.** 
  Function calls in the PHP VM are currently not inlined like most compilers would do. This means, in some scenarios, the overhead can become quite significant. For example, let's say you have a `Sprite` class with `getPosition`, `getType`, etc. The operations executed with the data retrieved from those functions are often quite simple and will often be cheaper than the overhead of the function call to retrieve the data. So, just use properties directly for DTOs, components, and state objects.


To come back to your question :)

It's not a conscious decision, and I think we could adapt the directive quite easily, as most things should already be properly typed.

**Reply by AlwinGarside** [reply]
Thanks you very much for your elaborate answer! Very interesting.

I remember back when strict types were first introduced in PHP 7, it actually used to give a slight performance penalty due to the additional runtime check (at least, as long as no type coercion would've been done had `strict_types` been left disabled).
This always sounded like something the PHP core devs would eventually iron out, so I wouldn't be surprised if, at this point, strict types do indeed give a slight performance bump.

I guess the best approach would be to just benchmark it. I'm still unfamiliar with php-glfw and VISU, so is there perhaps a good benchmark you could point me at?
